File: NeetCode/46/46.py
# Definition for a binary tree node.
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:

        if not root:
            return root

        def _invertTree(root: TreeNode) -> TreeNode:

            tmp = root.right
            root.right = root.left
            root.left = tmp

            if root.right:
                _invertTree(root.right)
            if root.left:
                _invertTree(root.left)

            return root

        return _invertTree(root)

File: NeetCode/46/46.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    TreeNode* _invertTree(TreeNode* root){

        TreeNode* tmp = root->right;
        root->right = root->left;
        root->left = tmp;

        if(root->left){
            _invertTree(root->left);
        }

        if(root->right){
            _invertTree(root->right);
        }

        return root;
    }
public:
    TreeNode* invertTree(TreeNode* root) {
        if(root == nullptr){
            return root;
        }

        return _invertTree(root);
    }
};
